cloud/scope/powervs_machine_test.go

```diff
@@ -122,74 +133,1188 @@ func newDHCPServer(serverID, networkID string) models.DHCPServers {
+			expectedServiceInstanceID: "service-instance-2",
+			machineScope: PowerVSMachineScope{
+				IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{
+					Spec: infrav1beta2.IBMPowerVSClusterSpec{
+						ServiceInstance: &infrav1beta2.IBMPowerVSResourceReference{
+							ID: ptr.To("service-instance-2"),
+						},
+					},
+				},
+			},
+			expectedError: nil,
+		}, {
+			name:                      "get service instance ID with serviceInstanceID present in both IBMPowerVSCluster Status and Spec ",
+			expectedServiceInstanceID: "service-instance-in-status",
+			machineScope: PowerVSMachineScope{
+				IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{
+					Status: infrav1beta2.IBMPowerVSClusterStatus{
+						ServiceInstance: &infrav1beta2.ResourceReference{
+							ID: ptr.To("service-instance-in-status"),
+						},
+					},
+					Spec: infrav1beta2.IBMPowerVSClusterSpec{
+						ServiceInstanceID: "service-instance-in-spec",
+					},
+				},
+			},
+			expectedError: nil,
+		}, {
+			name:                      "Failed to find service instance id",
+			expectedServiceInstanceID: "",
+			machineScope: PowerVSMachineScope{
+				IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{
+					Spec: infrav1beta2.IBMPowerVSClusterSpec{
+						ServiceInstance: &infrav1beta2.IBMPowerVSResourceReference{},
+					},
+				},
+			},
+			expectedError: fmt.Errorf("failed to find service instance id as both name and id are not set"),
+		},
+	}
+
+	for _, tc := range testcases {
+		g := NewWithT(t)
+		t.Run(tc.name, func(_ *testing.T) {
+			serviceInstanceID, err := tc.machineScope.GetServiceInstanceID()
+			g.Expect(serviceInstanceID).To(Equal(tc.expectedServiceInstanceID))
+			if tc.expectedError != nil {
+				g.Expect(err).To(Equal(tc.expectedError))
+			} else {
+				g.Expect(err).To(BeNil())
+			}
+		})
+	}
+
+	// Running other test cases which need some mock calls to be defined
+	var mockCtrl *gomock.Controller
+
+	setup := func(t *testing.T) {
+		t.Helper()
+		mockCtrl = gomock.NewController(t)
+	}
+	teardown := func() {
+		mockCtrl.Finish()
+	}
+
+	mockResourceController := resourcecontrollermock.NewMockResourceController(gomock.NewController(t))
+	t.Run("Returns service instance ID successfully when name is set in spec", func(t *testing.T) {
+		g := NewWithT(t)
+		setup(t)
+		t.Cleanup(teardown)
+		scope := PowerVSMachineScope{
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{
+				Spec: infrav1beta2.IBMPowerVSClusterSpec{
+					ServiceInstance: &infrav1beta2.IBMPowerVSResourceReference{
+						Name: core.StringPtr("foo-cluster"),
+					},
+				},
+			},
+			IBMPowerVSMachine: &infrav1beta2.IBMPowerVSMachine{
+				Status: infrav1beta2.IBMPowerVSMachineStatus{
+					Zone: core.StringPtr("us-south-1"),
+				},
+			},
+		}
+		mockResourceController.EXPECT().GetServiceInstance("", "foo-cluster", gomock.Any()).Return(&resourcecontrollerv2.ResourceInstance{GUID: core.StringPtr("foo-id")}, nil)
```

Comment:
aligning with other files we can replace core.StringPtr() with  ptr.To("foo-id") to keep it uniform. (Wherever applicable)